### PR TITLE
server: write pebble log messages to storage channel

### LINKED ISSUE #72683: storage: stray pebble log entries on test startup
Informs #58938

**Describe the problem**

When initializing clusters with `TestCluster` in tests, I see pebble log message **without their context tags**:

```
INFO DEV (gostd) ingest.go:616 [-] ‹ingest failed to remove original file: invalid argument›
```

The use of the DEV channel and lack of logging tags (`[-]`) in that message tells me that this is not using the `pebbleLogger` set up by cockroachdb, as I'd be expecting at least the tags `pebble` and `s` in there:

```go
                                                                                                                                                                                                                                                          // The context dance here is done so that we have a clean context without                                                                                                                                                                          inlining call to logtags.FromContext
  // timeouts that has a copy of the log tags.       
  logCtx ≔ logtags.WithTags(context.Background(), logtags.FromContext(ctx))
  logCtx = logtags.AddTag(logCtx, "pebble", nil)
  // The store id, could not necessarily be determined when this function
  // is called. Therefore, we use a container for the store id.
  storeIDContainer ≔ &base.StoreIDContainer{}
  logCtx = logtags.AddTag(logCtx, "s", storeIDContainer)

  if cfg.Opts.Logger ≡ nil {
    cfg.Opts.Logger = pebbleLogger{
      ctx:   logCtx,
      depth: 1,
    }
  }
```

The problem is that the context tags are missing and the log channels is wrong: we'll want to have the right channel and the tag information in all logs when collecting logs over the network for troubleshooting.

**To Reproduce**

Initialize a `TestCluster` in a test, stare at the generated logs on the STORAGE channel.
(NB: these are usually emitted only to files in a test)

**Expected behavior**

All Pebble log messages have context tags attached.

**Environment:**

`master` branch



Jira issue: CRDB-11256

### LINKED ISSUE #90483: No crash reason in logs despite roachtest: tpcc/mixed-headroom/n5cpu16 failing due to process exiting with exit code 1
@renatolabs was debugging a roachtest failure and did the following the discover that the cause of the failure was https://github.com/cockroachdb/pebble/issues/2019:

> The logs didn’t include anything indeed, which is why this has been a tough test failure to debug (we need to fix whatever is not right with logging). I instrumented exit.WithCode as suggested in this thread but that led nowhere. To get the stack trace, I compiled cockroach with a patched version of the stdlib that dumps a stack trace on os.Exit, and running the test with that

This indicates a bug of some kind. We do not expect to have a process crash without any info about the crash reason in the logs, if the CRDB process is doing the crashing (as opposed to the OOM killer or similar). With that said, it might be a test-only bug.

Here's a theory: Within a roachtest context but NOT within a production context, we are using the pebble default logger, instead of the CRDB logger. The latter takes care to get a log line in an expected file; the former does not seem to, based on the presence of this bug. See the pebble default logger here: https://github.com/cockroachdb/pebble/blob/bee0c60e96bc1f12308484c9789c10cb4b19f77a/logger.go#L29. Within the context of https://github.com/cockroachdb/pebble/issues/2019, I believe it is called here: https://github.com/cockroachdb/pebble/blob/master/internal/base/filenames.go#L167. Actually, I think the stacktrace given at https://github.com/cockroachdb/pebble/issues/2019 confirms that within the context of the roachtest we are using the pebble default logger: `github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/logger.go:32 +0x49` is in the stacktrace, and that corresponds to this line in the code: https://github.com/cockroachdb/pebble/blob/bee0c60e96bc1f12308484c9789c10cb4b19f77a/logger.go#L32. With all that said, I am not sure *why* the pebble default logger fails to get a log line in an expected file pre-crash.

Jira issue: CRDB-20795

### PR DESCRIPTION
**Pull request by renatolabs**
Previously, when setting up the server, the pebble engine would be initialized with Pebble's default logger. The reason for this is that the pebble initialization code calls `EnsureDefaults` on the configuration options _before_ checking if the `options.Logger` is nil/unset. At that point, it will never be unset, as `EnsureDefaults` will set the logger to `pebble.DefaultLogger` if it was not previously set.

This change overwrites the pebble logger if its found to be the `DefaultLogger`. We never want to use pebble's `DefaultLogger` in CRDB as that would mean pebble would use the standard library `log` package, making every message emitted by Pebble to be treated as `INFO` level messages, regardless of severity (including `log.Fatal` calls).

Related to #83079.

Fixes #72683.
Fixes #90483.

Release note: None.

### CODE REVIEW

**Comment by renatolabs** [comment]
`cockroach.log` before:

```
```

(_nothing_)

`cockroach.log` after:

```
F221021 19:42:58.902679 625530 3@pebble/internal/base/filenames.go:167 ⋮ [n1,s1,pebble] 915  ‹13314777078855381451.sst›:
F221021 19:42:58.902679 625530 3@pebble/internal/base/filenames.go:167 ⋮ [n1,s1,pebble] 915 +open ‹/mnt/data1/cockroach/13314777078855381451.sst›: no such file or directory
F221021 19:42:58.902679 625530 3@pebble/internal/base/filenames.go:167 ⋮ [n1,s1,pebble] 915 +directory contains 2910 files, 13 unknown, 2889 tables, 3 logs, 2 manifests
F221021 19:42:58.902679 625530 3@pebble/internal/base/filenames.go:167 ⋮ [n1,s1,pebble] 915 !goroutine 625530 [running]:

<stracktrace dumps>
```

**Comment by renatolabs** [comment]
Another approach to fix the issue would be to move the log initialization code so that it sits before the `EnsureDefaults()` call. I chose to keep the log setup logic where it is because: 1. It stays near other `cfg.Opts`-related code; and 2. Makes it explicit we never want `pebble.DefaultLogger`.

That said, I'm open to moving the code around if that's deemed to be more appropriate.

**Comment by joshimhoff** [comment]
My initial thought was this was a test-only bug, but based on this bug fix, does this affect production CRDB too?

If it is NOT a test-only bug, then I bet it is a regression that only affects certain CRDB versions, etc., since we use pebble logs all the time in production. Is that correct? If it is, when was the regression introduce?

Also, want to mark this PR as fixing https://github.com/cockroachdb/pebble/issues/2039?

**Comment by renatolabs** [comment]
> does this affect production CRDB too?

Yes, it affects production CRDB too.

> only affects certain CRDB versions, etc., since we use pebble logs all the time in production. Is that correct?

It seems like this has been the case for some time. The `pebble.log` logs that you are probably familiar with are event logs; these are set to the appropriate channel:

https://github.com/cockroachdb/cockroach/blob/62175c94eb7f041dbead36f358c326a31f3e2688/pkg/storage/pebble.go#L918-L924

**Comment by joshimhoff** [comment]
> It seems like this has been the case for some time. The pebble.log logs that you are probably familiar with are event logs.

Ah! Nice. Great find then.

File: pkg/storage/pebble.go [hunk 1]
```diff
@@ -853,7 +853,11 @@ func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
 	logCtx = logtags.AddTag(logCtx, "s", storeIDContainer)
 	logCtx = logtags.AddTag(logCtx, "pebble", nil)
 
-	if cfg.Opts.Logger == nil {
+	// if no logger was passed, the previous call to `EnsureDefaults` on
```
Comment (knz) [inline pkg/storage/pebble.go:856]: nit: capital at beginning of sentence, period at the end of comment.

**Review by knz — APPROVED** [review]
nice :100:

**Review by knz — APPROVED** [review]
nice :100:

This might be the fix to #72683. Can you check?

**Comment by renatolabs** [comment]
> nice 💯
> 
> This might be the fix to #72683. Can you check?

This change does fix that issue. The log message reported there is now displayed as below. Commit and PR messages updated to reflect that.

```
I221025 15:46:11.127184 5156 3@pebble/ingest.go:763 ⋮ [n2,s2,pebble] 184  ingest failed to remove original file: invalid argument
I221025 15:46:11.127453 5156 3@pebble/ingest.go:763 ⋮ [n2,s2,pebble] 185  ingest failed to remove original file: invalid argument
I221025 15:46:11.127458 5156 3@pebble/ingest.go:763 ⋮ [n2,s2,pebble] 186  ingest failed to remove original file: invalid argument
```

TFTR!

**Comment by renatolabs** [comment]
bors r=knz

**Review by srosenberg — APPROVED** [review]

**Pull request merged**